src/libtriton/engines/solver/bitwuzla/bitwuzlaSolver.cpp

```diff
@@ -0,0 +1,245 @@
+//! \file
+/*
+**  Copyright (C) - Triton
+**
+**  This program is under the terms of the Apache License 2.0.
+*/
+
+#include <fstream>
+#include <regex>
+#include <string>
+
+#include <triton/astContext.hpp>
+#include <triton/bitwuzlaSolver.hpp>
+#include <triton/exceptions.hpp>
+#include <triton/solverModel.hpp>
+#include <triton/symbolicExpression.hpp>
+#include <triton/symbolicVariable.hpp>
+#include <triton/tritonToBitwuzlaAst.hpp>
+#include <triton/tritonTypes.hpp>
+
+
+namespace triton {
+  namespace engines {
+    namespace solver {
+
+      int32_t BitwuzlaSolver::terminate_fun(void* state) {
+        auto p = reinterpret_cast<SolverParams*>(state);
+
+        // Check time only once in every 1000 calls.
+        if (++p->call_cnt < 1000) {
+          return 0;
+        }
+        p->call_cnt = 0;
+
+        // Count delta.
+        auto delta = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - p->start).count();
+
+        // Check timeout expired.
+        if (p->timeout && delta > p->timeout) {
+          p->code = 1;
+          return 1;
+        }
+
+#if defined(__unix__)
+        // Check memory limit.
+        delta /= 1000;
+        if (p->memory_limit && delta > p->last_mem_check && delta % p->delay == 0) {
+          p->last_mem_check = delta;
+
+          size_t memory_usage = 0;
+          std::ifstream str("/proc/self/status");
+          if (str.is_open()) {
+            std::string line;
+            std::regex vmsize("VmRSS:\\s*([0-9]+)\\s*kB");
+            while (std::getline(str, line)) {
+              std::smatch match;
+              if (std::regex_match(line, match, vmsize)) {
+                memory_usage = strtoul(match.str(1).c_str(), NULL, 10) / 1024;
+                break;
+              }
+            }
+          }
+          if (!memory_usage) {
+            return 0;
+          }
+          if (memory_usage > p->memory_limit) {
+            p->code = 2;
+            return 2;
+          }
+          // The more memory is available the rarely we check current
+          // memory consumption in order to save resources.
+          if (memory_usage < p->memory_limit / 4) {
+            p->delay = 5;
+          }
+          else if (memory_usage < 3 * p->memory_limit / 4) {
+            p->delay = 2;
+          }
+          else {
+            p->delay = 1;
+          }
+        }
+#endif
+
+        return 0;
+      }
```

Comment:
Regarding this whole function. You return either 0, 1 or 2. It would not be better to use enums?

Reply:
We may use the already existing enum for solver status.

Reply:
Terminate function should return non-zero to terminate solving, zero otherwise. Refactored callback to return 0/1 and to use status_e for the reason of termination (`SolverParams.code`)